### PR TITLE
Change backbone finetuning strategy to allow for DDP

### PR DESCRIPTION
**Pull request by ksikka**
See lightning-ai/pytorch-lightning#20340 for the original issue. The implemented workaround no longer freezes/unfreezes layers via require_grad, instead we only modulate the learning rate of the backbone layer.

Unblocks #138

### CODE REVIEW

**Comment by ksikka** [comment]
image

Run with and without code changes:
Capture

File: lightning_pose/callbacks.py [hunk 1]
```diff
@@ -42,3 +42,61 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningMo
...
+        assert optimizer.param_groups[1]["name"].startswith("upsampling")
+
+        backbone_lr = optimizer.param_groups[0]["lr"]
+        upsampling_lr = optimizer.param_groups[1]["lr"]
+
+        optimizer.param_groups[0]["lr"] = self._get_next_backbone_lr(
+            pl_module.current_epoch, backbone_lr, upsampling_lr
+        )
+
+    def _get_next_backbone_lr(self, current_epoch, backbone_lr, upsampling_lr):
```
Comment (danbider) [inline lightning_pose/callbacks.py:86]: good to comment on the `if` conditions, what is this the LR schedule that is implemented here

File: lightning_pose/models/base.py [hunk 2]
```diff
@@ -589,39 +580,3 @@ def training_step(
         self.log("total_loss", total_loss, prog_bar=True)
 
         return {"loss": total_loss}
-
```
Comment (danbider) [inline lightning_pose/models/base.py:592]: good cleanup

File: tests/test_callbacks.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+def test_unfreeze_backbone():
+    unfreeze_backbone = UnfreezeBackbone(2)
+
+    # Test unfreezing at epoch 2.
+    assert unfreeze_backbone._get_next_backbone_lr(0, 0.0, 1e-3) == 0.0
+    assert unfreeze_backbone._get_next_backbone_lr(1, 0.0, 1e-3) == 0.0
+    assert unfreeze_backbone._get_next_backbone_lr(2, 0.0, 1e-3) == 1e-4
+
+    # Test warming up.
```
Comment (danbider) [inline tests/test_callbacks.py:11]: comment on the 1.5 factor?

File: tests/test_callbacks.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from lightning_pose.callbacks import UnfreezeBackbone
+
+def test_unfreeze_backbone():
+    unfreeze_backbone = UnfreezeBackbone(2)
+
+    # Test unfreezing at epoch 2.
```
Comment (danbider) [inline tests/test_callbacks.py:6]: comment that expected behavior is zero lr until epoch 2? then specify why 1e-4. which is the set LR?

**Review by danbider — APPROVED** [review]
very happy to see these updates -- thanks @ksikka!

**Review by themattinthehatt — APPROVED** [review]
